# 2. teden: Numpy - Domača naloga

Uvozimo modul numpy:

In [1]:
import numpy as np

### 1.Create a null vector of size 10

In [7]:
#test
print('Result_1 =', np.zeros(10))

Result_1 = [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [8]:
a=np.zeros(10)
a[4]=1
print('Result_2 =', a)

Result_2 = [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [19]:
a=np.linspace(10,49,40, dtype=int)
print('Result_3 =',a)

Result_3 = [10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]


Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [20]:
Z = np.arange(50)

In [21]:
a_invert=np.flip(Z)
print('Result_4 =',a_invert)

Result_4 = [49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26
 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2
  1  0]


Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [24]:
a_mat=np.arange(9).reshape(3,3)
print('Result_5 =',a_mat)
#vse ok, če je tipa int?

Result_5 = [[0 1 2]
 [3 4 5]
 [6 7 8]]


Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [62]:
# nevem zakaj ne izpisuje dtype, da je int64... je pa privzeto tako piše tudi v opisu. 
#https://docs.scipy.org/doc/numpy/reference/generated/numpy.nonzero.html

a = [1,2,0,0,4,0]


In [65]:
print('Result_6 =',np.nonzero(a))


Result_6 = (array([0, 1, 4]),)


Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [67]:
print('Resoult_7 =', np.identity(3))

Resoult_7 = [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [79]:
a=np.random.random((3,3,3))
print('Result_8 =',a)

Result_8 = [[[0.821501   0.67259027 0.33132733]
  [0.27953797 0.42175535 0.38027455]
  [0.72245969 0.51165252 0.61915146]]

 [[0.72101539 0.70186131 0.15636683]
  [0.26333059 0.64673534 0.99453936]
  [0.09760875 0.06335352 0.06795323]]

 [[0.96238817 0.24318663 0.16524251]
  [0.98014163 0.78478207 0.13209672]
  [0.62099719 0.15156849 0.44127062]]]


Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [85]:
a=np.random.rand(10,10)
print('Result_9_max =', a.max(),'Result_9_min =',a.min())

Result_9_max = 0.9937862753051339 Result_9_min = 0.003987043965719805


Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [95]:
x = np.ones((10,10))
x[1:-1,1:-1] = 0
print('Result_10 =', x)

Result_10 = [[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

Ne razumem čisto zakaj se gre, sem pa zagnal vsako vrstico...

In [88]:
0 * np.nan

nan

In [89]:
np.nan == np.nan

False

In [90]:
np.inf > np.nan

False

In [91]:
np.nan - np.nan

nan

In [92]:
0.3 == 3 * 0.1 

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [94]:
x = np.zeros((8,8),dtype=int)
x[1::2,::2] = 1
x[::2,1::2] = 1
print('Result_12 =',x)

Result_12 = [[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [104]:
Z = np.arange(11)

In [105]:
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [110]:
Z[(3 < Z) & (Z <= 8)] *= -1
print('Result_13 =',Z)

Result_13 = [ 0  1  2  3 -4 -5 -6 -7 -8  9 10]


Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [111]:
print('Result_14 =', np.linspace(0,1,11,endpoint=False)[1:])


Result_14 = [0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455
 0.63636364 0.72727273 0.81818182 0.90909091]


Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [122]:
a=np.random.rand(10)

In [123]:
print('Result_15 =',np.sort(a))

Result_15 = [0.02256497 0.3636466  0.39486468 0.67111998 0.7583439  0.78037263
 0.80828066 0.80893761 0.82852012 0.85295472]


Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [133]:
A = np.random.randint(0,2,5)
A

array([0, 1, 0, 0, 0])

In [134]:
B = np.random.randint(0,2,5)
B

array([1, 1, 1, 1, 0])

In [135]:
#primerjava po elementih
A==B

array([False,  True, False, False,  True])

In [136]:
#ker se ne ujemjo vsi elementi je rezultat False
min(A==B)

False

### 17.Make an array immutable (read-only)

In [137]:
Z = np.zeros(10)

In [138]:
Z.flags.writeable = False

In [139]:
#če probamo spremeniti katero od vrednosti nam ne pusti, kar pomeni, da je readonley
Z[1]=1

ValueError: assignment destination is read-only

### 18.Create random vector of size 10 and replace the maximum value by 0

In [155]:
a=np.random.rand(10)
a


array([0.38039154, 0.03062336, 0.59761421, 0.88701407, 0.67841997,
       0.05977271, 0.60352911, 0.01546868, 0.86148107, 0.94148078])

In [156]:
a.max()

0.9414807775304825

In [157]:
a[a==a.max()]=0
print('Result_18 =',a)


Result_18 = [0.38039154 0.03062336 0.59761421 0.88701407 0.67841997 0.05977271
 0.60352911 0.01546868 0.86148107 0.        ]


### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [168]:
Z = np.arange(100)
Z

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [169]:
v= np.random.uniform(0,100)
v

87.5672963311716

In [170]:
idx = (np.abs(Z-v)).argmin()

In [171]:
print('Result_19 =', Z[idx])

Result_19 = 88


### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [172]:
Z = np.arange(10, dtype=np.int32)

In [173]:
print(Z)

[0 1 2 3 4 5 6 7 8 9]


In [174]:
print('Result_20 =', Z.astype('float32'))

Result_20 = [0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]


Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

In [178]:
Z = np.random.randint(0,10,50)
print(Z)

[7 3 8 8 2 5 6 9 5 5 4 3 7 8 0 6 3 2 5 0 7 3 6 5 9 7 0 5 9 7 7 2 1 3 0 3 5
 1 8 0 8 4 3 0 4 2 4 9 4 9]


In [179]:
print('Result_21 =',np.bincount(Z).argmax())

Result_21 = 3
